```
======================================================================

ERROR: corehq.apps.users.tests.test_views:TestMobileWorkerListView.test_create_anonymous_mobile_worker

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/vendor/lib/python2.7/site-packages/mock.py", line 1224, in patched

    return func(*args, **keywargs)

  File "/mnt/commcare-hq-ro/corehq/apps/users/tests/test_views.py", line 67, in test_create_anonymous_mobile_worker

    'is_anonymous': True,

  File "/mnt/commcare-hq-ro/corehq/apps/users/tests/test_views.py", line 37, in _remote_invoke

    HTTP_X_REQUESTED_WITH='XMLHttpRequest',

  File "/vendor/lib/python2.7/site-packages/django/test/client.py", line 541, in post

    secure=secure, **extra)

  File "/vendor/lib/python2.7/site-packages/django/test/client.py", line 343, in post

    secure=secure, **extra)

  File "/vendor/lib/python2.7/site-packages/django/test/client.py", line 409, in generic

    return self.request(**r)

  File "/vendor/lib/python2.7/site-packages/django/test/client.py", line 494, in request

    six.reraise(*exc_info)

  File "/vendor/lib/python2.7/site-packages/django/core/handlers/exception.py", line 42, in inner

    response = get_response(request)

  File "/vendor/lib/python2.7/site-packages/django/core/handlers/base.py", line 187, in _get_response

    response = self.process_exception_by_middleware(e, request)

  File "/vendor/lib/python2.7/site-packages/django/core/handlers/base.py", line 185, in _get_response

    response = wrapped_callback(request, *callback_args, **callback_kwargs)

  File "/vendor/lib/python2.7/site-packages/django/views/generic/base.py", line 68, in view

    return self.dispatch(request, *args, **kwargs)

  File "/mnt/commcare-hq-ro/corehq/apps/style/decorators.py", line 17, in _wrapped

    return view_func(class_based_view, request, *args, **kwargs)

  File "/mnt/commcare-hq-ro/corehq/apps/style/decorators.py", line 53, in _wrapped

    return view_func(class_based_view, request, *args, **kwargs)

  File "/vendor/lib/python2.7/site-packages/django/utils/decorators.py", line 67, in _wrapper

    return bound_func(*args, **kwargs)

  File "/mnt/commcare-hq-ro/corehq/apps/domain/decorators.py", line 103, in _inner

    return view_func(req, domain_name, *args, **kwargs)

  File "/mnt/commcare-hq-ro/corehq/apps/users/decorators.py", line 21, in _inner

    return view_func(request, domain, *args, **kwargs)

  File "/vendor/lib/python2.7/site-packages/django/utils/decorators.py", line 63, in bound_func

    return func.__get__(self, type(self))(*args2, **kwargs2)

  File "/mnt/commcare-hq-ro/corehq/apps/users/views/mobile/users.py", line 515, in dispatch

    return super(MobileWorkerListView, self).dispatch(*args, **kwargs)

  File "/vendor/lib/python2.7/site-packages/django/utils/decorators.py", line 67, in _wrapper

    return bound_func(*args, **kwargs)

  File "/mnt/commcare-hq-ro/corehq/apps/domain/decorators.py", line 103, in _inner

    return view_func(req, domain_name, *args, **kwargs)

  File "/vendor/lib/python2.7/site-packages/django/utils/decorators.py", line 63, in bound_func

    return func.__get__(self, type(self))(*args2, **kwargs2)

  File "/mnt/commcare-hq-ro/corehq/apps/domain/views.py", line 238, in dispatch

    return super(LoginAndDomainMixin, self).dispatch(*args, **kwargs)

  File "/vendor/lib/python2.7/site-packages/django/views/generic/base.py", line 88, in dispatch

    return handler(request, *args, **kwargs)

  File "/vendor/lib/python2.7/site-packages/djangular/views/mixins.py", line 102, in post

    response_data = handler(in_data)

  File "/mnt/commcare-hq-ro/corehq/apps/users/views/mobile/users.py", line 736, in create_mobile_worker

    couch_user = self._build_anonymous_commcare_user()

  File "/mnt/commcare-hq-ro/corehq/apps/users/views/mobile/users.py", line 767, in _build_anonymous_commcare_user

    location=SQLLocation.objects.get(location_id=location_id),

  File "/vendor/lib/python2.7/site-packages/django/db/models/manager.py", line 85, in manager_method

    return getattr(self.get_queryset(), name)(*args, **kwargs)

  File "/vendor/lib/python2.7/site-packages/django/db/models/query.py", line 385, in get

    self.model._meta.object_name

DoesNotExist: SQLLocation matching query does not exist.

-------------------- >> begin captured logging << --------------------

auditcare.middleware: DEBUG: Middleware is running in a test context, disabling monkeypatch

--------------------- >> end captured logging << ---------------------
```

--- a/corehq/apps/users/signals.py
+++ b/corehq/apps/users/signals.py
@@ -8,10 +8,6 @@ from corehq.elastic import send_to_elasticsearch
 commcare_user_post_save = Signal(providing_args=["couch_user"])
 couch_user_post_save = Signal(providing_args=["couch_user"])
 
-# Called after user validation, before save.
-# Used for additional validation or modification.
-clean_commcare_user = Signal(providing_args=["domain", "request_user", "user", "forms"])
-
 
 @receiver(user_logged_in)
 def set_language(sender, **kwargs):

--- a/corehq/apps/users/views/__init__.py
+++ b/corehq/apps/users/views/__init__.py
@@ -73,7 +73,6 @@ from corehq.apps.users.forms import (
 from corehq.apps.users.models import (CouchUser, CommCareUser, WebUser, DomainRequest,
                                       DomainRemovalRecord, UserRole, AdminUserRole, Invitation,
                                       DomainMembershipError)
-from corehq.apps.users.signals import clean_commcare_user
 from corehq.elastic import ADD_TO_ES_FILTER, es_query
 from corehq.util.couch import get_document_or_404
 from corehq import toggles
@@ -277,29 +276,9 @@ class BaseEditUserView(BaseUserSettingsView):
         saved = False
         if self.request.POST['form_type'] == "commtrack":
             if self.commtrack_form.is_valid():
-                clean_commcare_user.send(
-                    'BaseEditUserView.commtrack',
-                    domain=self.domain,
-                    request_user=self.request.couch_user,
-                    user=self.editable_user,
-                    forms={self.commtrack_form.__class__.__name__: self.commtrack_form}
-                )
-                if self.commtrack_form.is_valid():
-                    self.commtrack_form.save(self.editable_user)
-                    saved = True
+                self.commtrack_form.save(self.editable_user)
+                saved = True
         elif self.request.POST['form_type'] == "update-user":
-            self.form_user_update.is_valid()
-            forms = {self.form_user_update.__class__.__name__: self.form_user_update}
-            if hasattr(self, 'custom_data'):
-                self.custom_data.is_valid()
-                forms[self.custom_data.__class__.__name__] = self.custom_data
-            clean_commcare_user.send(
-                'BaseEditUserView.update_user',
-                domain=self.domain,
-                request_user=self.request.couch_user,
-                user=self.editable_user,
-                forms=forms
-            )
             if all([self.update_user(), self.custom_user_is_valid()]):
                 messages.success(self.request, _('Changes saved for user "%s"') % self.editable_user.raw_username)
                 saved = True

--- a/corehq/apps/users/views/mobile/users.py
+++ b/corehq/apps/users/views/mobile/users.py
@@ -754,10 +754,7 @@ class MobileWorkerListView(JSONResponseMixin, BaseUserSettingsView):
             if not is_valid():
                 return {'error': _("Forms did not validate")}
 
-            couch_user.save()
-            location_id = self.new_mobile_worker_form.cleaned_data['location_id']
-            if location_id:
-                couch_user.set_location(SQLLocation.objects.get(location_id=location_id))
+            couch_user = self._build_commcare_user()
 
         return {
             'success': True,
@@ -774,7 +771,7 @@ class MobileWorkerListView(JSONResponseMixin, BaseUserSettingsView):
         last_name = ANONYMOUS_LASTNAME
         location_id = self.new_anonymous_mobile_worker_form.cleaned_data['location_id']
 
-        couch_user = CommCareUser.create(
+        return CommCareUser.create(
             self.domain,
             format_username(username, self.domain),
             password,
@@ -783,16 +780,15 @@ class MobileWorkerListView(JSONResponseMixin, BaseUserSettingsView):
             last_name=last_name,
             user_data=self.custom_data.get_data_to_save(),
             is_anonymous=True,
+            location=SQLLocation.objects.get(location_id=location_id),
         )
-        if location_id:
-            couch_user.set_location(SQLLocation.objects.get(location_id=location_id))
-        return couch_user
 
     def _build_commcare_user(self):
         username = self.new_mobile_worker_form.cleaned_data['username']
         password = self.new_mobile_worker_form.cleaned_data['password']
         first_name = self.new_mobile_worker_form.cleaned_data['first_name']
         last_name = self.new_mobile_worker_form.cleaned_data['last_name']
+        location_id = self.new_mobile_worker_form.cleaned_data['location_id']
 
         return CommCareUser.create(
             self.domain,
@@ -802,7 +798,7 @@ class MobileWorkerListView(JSONResponseMixin, BaseUserSettingsView):
             first_name=first_name,
             last_name=last_name,
             user_data=self.custom_data.get_data_to_save(),
-            commit=False,
+            location=SQLLocation.objects.get(location_id=location_id),
         )
 
     def _ensure_proper_request(self, in_data):

--- a/corehq/apps/users/views/mobile/users.py
+++ b/corehq/apps/users/views/mobile/users.py
@@ -764,7 +764,7 @@ class MobileWorkerListView(JSONResponseMixin, BaseUserSettingsView):
             last_name=last_name,
             user_data=self.custom_data.get_data_to_save(),
             is_anonymous=True,
-            location=SQLLocation.objects.get(location_id=location_id),
+            location=SQLLocation.objects.get(location_id=location_id) if location_id else None,
         )
 
     def _build_commcare_user(self):
@@ -782,7 +782,7 @@ class MobileWorkerListView(JSONResponseMixin, BaseUserSettingsView):
             first_name=first_name,
             last_name=last_name,
             user_data=self.custom_data.get_data_to_save(),
-            location=SQLLocation.objects.get(location_id=location_id),
+            location=SQLLocation.objects.get(location_id=location_id) if location_id else None,
         )
 
     def _ensure_proper_request(self, in_data):

--- a/corehq/apps/users/views/mobile/users.py
+++ b/corehq/apps/users/views/mobile/users.py
@@ -74,7 +74,6 @@ from corehq.apps.users.forms import (
 )
 from corehq.apps.users.models import CommCareUser, CouchUser
 from corehq.apps.users.const import ANONYMOUS_USERNAME, ANONYMOUS_FIRSTNAME, ANONYMOUS_LASTNAME
-from corehq.apps.users.signals import clean_commcare_user
 from corehq.apps.users.tasks import bulk_upload_async, turn_on_demo_mode_task, reset_demo_user_restore_task, \
     bulk_download_users_async
 from corehq.apps.users.util import can_add_extra_mobile_workers, format_username
@@ -277,13 +276,6 @@ class EditCommCareUserView(BaseEditUserView):
             phone_number = self.request.POST['phone_number']
             phone_number = re.sub('\s', '', phone_number)
             if re.match(r'\d+$', phone_number):
-                clean_commcare_user.send(
-                    'EditCommCareUserView.phone_number',
-                    domain=self.domain,
-                    request_user=self.request.couch_user,
-                    user=self.editable_user,
-                    forms={'phone_number': phone_number},
-                )
                 self.editable_user.add_phone_number(phone_number)
                 self.editable_user.save()
                 messages.success(request, _("Phone number added!"))
@@ -759,9 +751,6 @@ class MobileWorkerListView(JSONResponseMixin, BaseUserSettingsView):
                 return {'error': _("Forms did not validate")}
             couch_user = self._build_anonymous_commcare_user()
         else:
-            is_valid()
-            couch_user = self._build_commcare_user()
-            self.send_clean_commcare_user_signal(couch_user)
             if not is_valid():
                 return {'error': _("Forms did not validate")}
 
@@ -777,18 +766,6 @@ class MobileWorkerListView(JSONResponseMixin, BaseUserSettingsView):
                 args=[self.domain, couch_user.userID]
             )
         }
-
-    def send_clean_commcare_user_signal(self, couch_user):
-        clean_commcare_user.send(
-            'MobileWorkerListView.create_mobile_worker',
-            domain=self.domain,
-            request_user=self.request.couch_user,
-            user=couch_user,
-            forms={
-                self._mobile_worker_form.__class__.__name__: self._mobile_worker_form,
-                self.custom_data.__class__.__name__: self.custom_data,
-            }
-        )
 
     def _build_anonymous_commcare_user(self):
         username = ANONYMOUS_USERNAME

--- a/custom/enikshay/tests/test_user_setup.py
+++ b/custom/enikshay/tests/test_user_setup.py
@@ -10,7 +10,6 @@ from corehq.apps.locations.views import LocationFieldsView
 from corehq.apps.users.models import CommCareUser, WebUser, UserRole, Permissions
 from corehq.apps.users.views.mobile.custom_data_fields import CUSTOM_USER_DATA_FIELD_TYPE, UserFieldsView
 from corehq.apps.users.forms import UpdateCommCareUserInfoForm
-from corehq.apps.users.signals import clean_commcare_user
 from corehq.apps.users.util import format_username
 from .utils import setup_enikshay_locations
 from ..user_setup import (
@@ -187,14 +186,7 @@ class TestUserSetupUtils(TestCase):
         )
         self.assertValid(user_form)
         self.assertValid(custom_data)
-        clean_commcare_user.send(
-            'BaseEditUserView.update_user',
-            domain=self.domain,
-            request_user=self.web_user,
-            user=user,
-            forms={'UpdateCommCareUserInfoForm': user_form,
-                   'CustomDataEditor': custom_data}
-        )
+        # TODO update this test to account for form subclasses
         self.assertValid(user_form)
         self.assertInvalid(custom_data)  # there should be an error
 

--- a/custom/enikshay/user_setup.py
+++ b/custom/enikshay/user_setup.py
@@ -18,7 +18,7 @@ from corehq.apps.custom_data_fields import CustomDataEditor
 from corehq.apps.locations.forms import LocationFormSet, LocationForm
 from corehq.apps.users.forms import NewMobileWorkerForm, clean_mobile_worker_username
 from corehq.apps.users.models import CommCareUser
-from corehq.apps.users.signals import clean_commcare_user, commcare_user_post_save
+from corehq.apps.users.signals import commcare_user_post_save
 from .const import AGENCY_USER_FIELDS, AGENCY_LOCATION_FIELDS
 from .models import IssuerId
 
@@ -261,7 +261,6 @@ def add_drtb_hiv_to_dto(domain, user):
 
 
 def connect_signals():
-    clean_commcare_user.connect(clean_user_callback, dispatch_uid="clean_user_callback")
     commcare_user_post_save.connect(save_user_callback, dispatch_uid="save_user_callback")
 
 

